File: src/algorithmic_toolbox/week3/MaxLoot.java
package algorithmic_toolbox.week3;

import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

class Loot {
  private int value;
  private int mass;

  public static final Comparator<Loot> BY_UNIT_VALUE = (
    Comparator.comparingDouble((Loot loot) -> loot.calcValue(1))
  );

  public Loot(int value, int mass) {
    this.value = value;
    this.mass = mass;
  }

  public int getMass() {
    return mass;
  }

  public int getValue() {
    return value;
  }

  public double calcValue(int mass) {
    return (double) value / this.mass * mass;
  }
}

public class MaxLoot {

  public static double maxValue(Loot[] items, int capacity) {
    double totalValue = 0;
    for (Loot loot : items) {
      if (capacity >= loot.getMass()) {
        totalValue += loot.getValue();
        capacity -= loot.getMass();
      } else {
        totalValue += loot.calcValue(capacity);
        capacity = 0;
        break;
      }
    }

    return totalValue;
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int numItems = scanner.nextInt();
    int capacity = scanner.nextInt();

    Loot[] items = new Loot[numItems];
    for (int i = 0; i < numItems; i++) {
      items[i] = new Loot(scanner.nextInt(), scanner.nextInt());
    }
    scanner.close();

    Arrays.sort(items, Loot.BY_UNIT_VALUE.reversed());
    double maxValue = maxValue(items, capacity);
    System.out.printf("%.4f\n", maxValue);
  }
}

File: src/algorithmic_toolbox/week3/max_loot.py
from typing import List


class Loot:
    __slots__ = ['_mass', '_total_value']

    def __init__(self, total_value: int, mass: int):
        self._mass = mass
        self._total_value = total_value

    @property
    def mass(self) -> int:
        return self._mass

    @property
    def value(self) -> int:
        return self._total_value

    def calc_value(self, mass: int) -> float:
        return self.value / self.mass * mass

    def __repr__(self) -> str:
        price_per_unit = round(self.value / self.mass, 4)
        return f'Loot({self.mass} units @ {price_per_unit}/unit)'


if __name__ == '__main__':
    num_items, capacity = map(int, input().split())

    items: List[Loot] = []
    for _ in range(num_items):
        items.append(Loot(*map(int, input().split())))

    items.sort(key=lambda loot: loot.value / loot.mass, reverse=True)

    total_value = 0
    for loot in items:
        if capacity >= loot.mass:
            capacity -= loot.mass
            total_value += loot.value
        else:
            total_value += loot.calc_value(capacity)
            capacity = 0
            break

    print(round(float(total_value), 4))
